staging-service-shared/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/BatchPathProvider.java

```diff
@@ -18,20 +18,38 @@
 import com.github.cafdataprocessing.services.staging.BatchId;
 import com.github.cafdataprocessing.services.staging.TenantId;
 import com.github.cafdataprocessing.services.staging.exceptions.StagingException;
+import com.github.cafdataprocessing.services.staging.utils.ServiceIdentifier;
+import com.github.cafdataprocessing.services.staging.utils.Tracker;
 import java.io.File;
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class BatchPathProvider
 {
     private static final Logger LOGGER = LoggerFactory.getLogger(BatchPathProvider.class);
-    private static final String INPROGRESS_FOLDER = "in_progress";
+    public static final String INPROGRESS_FOLDER = "in_progress";
     public static final String COMPLETED_FOLDER = "completed";
 
+    public static final String MOVED_TO_PROGRESS = "The in-progress folder might have been moved to Completed.";
     private final Path basePath;
 
+    private final List<Path> inProgressBatchesOfDifferentServices = new ArrayList<>();
```

Comment:
This has been introduced since the last code review and its introduction is a disaster.  It has severely damaged the code base by attempting to operate through a side-effect.  Furthermore you only ever add to it so it is both a memory leak and a performance issue as the number of paths build up.

Reply:
**Resolved in this commit:**
[590f331c0db34ae520d5a2d8a635c62a07f8bd09] Code review: Extensive refactoring

I scrapped this variable and updated the method (which is now called `getOtherServicesInProgressBatchDirectories()`) to directly return the list of paths which correspond to other services which are uploading the batch.  There is no need to also return the paths that correspond to uploads being done by this service since they can be looked up in the concurrent map object.